File: src/lib/finance/load-calculator.ts
/**
 * LOAD CALCULATOR
 * ===============
 *
 * Two questions, one cost model.
 *
 *   EVALUATE   "Broker is offering $700 for this. Is it worth running?"
 *   TARGET     "I want $1.50/mile of profit. What do I quote?"
 *
 * The cost model, in the order money leaves:
 *
 *   Total miles      = loaded + deadhead to pickup          (always both)
 *   Fuel             = total miles / MPG x price per gallon
 *   Tolls            = entered
 *   Dispatch         = percent of gross, or a flat amount
 *   Factoring        = percent of gross, or a flat amount
 *   Other            = entered
 *   Allocated ops    = total miles x normalized operating cost per mile
 *   Debt burden      = total miles x debt-service cost per mile (cash only)
 *
 * `overheadCostPerMile` is the truck's own historical cost per mile with
 * fuel, tolls, dispatch and factoring removed -- those four are entered
 * explicitly above, and charging them twice is the classic way a load
 * calculator lies to you. What is left is insurance, parking, permits,
 * maintenance, repairs and the rest of the operating overhead every mile has
 * to carry. Truck-note Debt Service is excluded from operating overhead and
 * displayed in its own cash-burden layer.
 *
 * Deadhead is never optional. A rate is only as good as the miles it took to
 * earn it, and the empty ones are miles.
 */

import { div, roundMoney, type RatingThresholds } from "../calculations";
import { calculateLoadScore, type LoadScore } from "./load-score";

export type FeeMode = "PCT" | "AMOUNT";

export interface LoadEstimateInput {
  grossRate: number;
  loadedMiles: number;
  deadheadMiles: number;
  fuelPrice: number;
  mpg: number;
  tolls: number;
  dispatchMode: FeeMode;
  dispatchValue: number;
  factoringMode: FeeMode;
  factoringValue: number;
  otherCost: number;
  /** Historical overhead per mile. See `overheadCostPerMile`. */
  overheadPerMile: number;
  /** Historical debt-service cash burden; never used by the load rating. */
  debtServicePerMile?: number;
}

export interface CostLineItem {
  key: string;
  label: string;
  amount: number;
  note?: string;
}

export interface LoadEstimate {
  totalMiles: number;
  grossPerLoadedMile: number;
  grossPerTotalMile: number;
  deadheadPct: number;
  gallons: number;
  fuelCost: number;
  tolls: number;
  dispatch: number;
  factoring: number;
  otherCost: number;
  overhead: number;
  debtService: number;
  /** Fuel + tolls + dispatch + factoring + other. What the trip itself costs. */
  tripCost: number;
  directTripCosts: number;
  contributionProfit: number;
  contributionProfitPerMile: number;
  contributionMargin: number;
  allocatedOperatingCosts: number;
  fullyLoadedOperatingProfit: number;
  fullyLoadedOperatingProfitPerMile: number;
  fullyLoadedOperatingMargin: number;
  cashAfterDebtService: number;
  /** Trip cost plus the overhead those miles have to carry. */
  totalCost: number;
  profit: number;
  profitPerMile: number;
  profitMargin: number;
  costPerMile: number;
  breakEvenRate: number;
  operatingBreakEven: number;
  cashBreakEven: number;
  lines: CostLineItem[];
  score: LoadScore;
  /** False when miles or MPG are missing: nothing below can be trusted yet. */
  valid: boolean;
}

function feeAmount(mode: FeeMode, value: number, gross: number): number {
  const safe = Number.isFinite(value) && value > 0 ? value : 0;
  return mode === "PCT" ? roundMoney(gross * (safe / 100)) : roundMoney(safe);
}

export function calculateLoadEstimate(
  input: LoadEstimateInput,
  thresholds: RatingThresholds,
  deadheadWarnPct: number,
): LoadEstimate {
  const loadedMiles = Math.max(0, input.loadedMiles);
  const deadheadMiles = Math.max(0, input.deadheadMiles);
  const totalMiles = loadedMiles + deadheadMiles;
  const grossRate = Math.max(0, input.grossRate);

  const gallons = input.mpg > 0 ? totalMiles / input.mpg : 0;
  const fuelCost = roundMoney(gallons * Math.max(0, input.fuelPrice));
  const tolls = roundMoney(Math.max(0, input.tolls));
  const dispatch = feeAmount(input.dispatchMode, input.dispatchValue, grossRate);
  const factoring = feeAmount(input.factoringMode, input.factoringValue, grossRate);
  const otherCost = roundMoney(Math.max(0, input.otherCost));
  const overhead = roundMoney(totalMiles * Math.max(0, input.overheadPerMile));
  const debtService = roundMoney(totalMiles * Math.max(0, input.debtServicePerMile ?? 0));

  const tripCost = roundMoney(fuelCost + tolls + dispatch + factoring + otherCost);
  const totalCost = roundMoney(tripCost + overhead);
  const contributionProfit = roundMoney(grossRate - tripCost);
  const contributionProfitPerMile = div(contributionProfit, totalMiles);
  const contributionMargin = div(contributionProfit, grossRate) * 100;
  const fullyLoadedOperatingProfit = roundMoney(contributionProfit - overhead);
  const fullyLoadedOperatingProfitPerMile = div(fullyLoadedOperatingProfit, totalMiles);
  const fullyLoadedOperatingMargin = div(fullyLoadedOperatingProfit, grossRate) * 100;
  const cashAfterDebtService = roundMoney(fullyLoadedOperatingProfit - debtService);
  const deadheadPct = div(deadheadMiles, totalMiles) * 100;

  const lines: CostLineItem[] = [
    {
      key: "fuel",
      label: "Fuel",
      amount: fuelCost,
      note:
        input.mpg > 0
          ? `${gallons.toFixed(1)} gal at $${input.fuelPrice.toFixed(2)}/gal, ${input.mpg.toFixed(1)} MPG`
          : "Enter MPG to estimate fuel",
    },
    { key: "tolls", label: "Tolls", amount: tolls },
    {
      key: "dispatch",
      label: "Dispatch",
      amount: dispatch,
      note: input.dispatchMode === "PCT" ? `${input.dispatchValue}% of gross` : "Flat fee",
    },
    {
      key: "factoring",
      label: "Factoring",
      amount: factoring,
      note: input.factoringMode === "PCT" ? `${input.factoringValue}% of gross` : "Flat fee",
    },
    { key: "other", label: "Other costs", amount: otherCost },
  ].filter((line) => line.amount > 0 || line.key === "fuel");

  return {
    totalMiles,
    grossPerLoadedMile: div(grossRate, loadedMiles),
    grossPerTotalMile: div(grossRate, totalMiles),
    deadheadPct,
    gallons,
    fuelCost,
    tolls,
    dispatch,
    factoring,
    otherCost,
    overhead,
    debtService,
    tripCost,
    directTripCosts: tripCost,
    contributionProfit,
    contributionProfitPerMile,
    contributionMargin,
    allocatedOperatingCosts: overhead,
    fullyLoadedOperatingProfit,
    fullyLoadedOperatingProfitPerMile,
    fullyLoadedOperatingMargin,
    cashAfterDebtService,
    totalCost,
    profit: fullyLoadedOperatingProfit,
    profitPerMile: fullyLoadedOperatingProfitPerMile,
    profitMargin: fullyLoadedOperatingMargin,
    costPerMile: div(totalCost, totalMiles),
    breakEvenRate: totalCost,
    operatingBreakEven: totalCost,
    cashBreakEven: roundMoney(totalCost + debtService),
    lines,
    score: calculateLoadScore(
      {
        profitPerMile: contributionProfitPerMile,
        profitMargin: contributionMargin,
        deadheadPct,
      },
      thresholds,
      deadheadWarnPct,
    ),
    valid: totalMiles > 0 && input.mpg > 0,
  };
}

/* ---- What rate should I ask? ------------------------------------------ */

export interface TargetRateInput {
  loadedMiles: number;
  deadheadMiles: number;
  fuelPrice: number;
  mpg: number;
  tolls: number;
  dispatchMode: FeeMode;
  dispatchValue: number;
  factoringMode: FeeMode;
  factoringValue: number;
  otherCost: number;
  overheadPerMile: number;
  debtServicePerMile?: number;
  /** What the owner wants to clear, per total mile. */
  targetProfitPerMile: number;
}

export interface RateTier {
  key: "operatingBreakeven" | "cashBreakeven" | "minimum" | "good" | "great" | "target";
  label: string;
  profitPerMile: number;
  rate: number;
  ratePerLoadedMile: number;
  description: string;
}

export interface TargetRate {
  totalMiles: number;
  gallons: number;
  fuelCost: number;
  tolls: number;
  otherCost: number;
  overhead: number;
  debtService: number;
  /** Costs that do not move with the rate. Dispatch and factoring do. */
  fixedTripCost: number;
  /** Direct costs that do not move with the rate, before allocated overhead. */
  directFixedCost: number;
  costPerMile: number;
  /** Combined dispatch + factoring share of gross, as a fraction (0.075). */
  grossFeeRate: number;
  /** Dispatch + factoring entered as flat dollars rather than a percentage. */
  flatFees: number;
  /** Direct trip costs only. Always available when the trip inputs are valid. */
  directCostBreakEven: number;
  tiers: RateTier[];
  valid: boolean;
  /** Set when fees are configured at 100% or more and no rate can work. */
  impossible: boolean;
}

export type OfferPosition = "GREAT" | "GOOD" | "MARGINAL" | "BELOW_MINIMUM";

export interface OfferComparison {
  position: OfferPosition;
  differenceVsGreat: number;
  /** The threshold the negotiation should aim to settle at. */
  settlementTarget: number | null;
  /** Null when the existing offer already meets or exceeds GREAT. */
  suggestedCounteroffer: number | null;
}

export const QUOTE_ROUNDING_INCREMENT = 25;
export const QUOTE_CUSHION_PCT = 0.03;
export const MIN_QUOTE_CUSHION = 25;

/**
 * A deterministic opening anchor: add 3% (at least $25), then round UP to
 * the next $25. The current offer is a hard floor and can never be rounded
 * down by the suggestion.
 */
export function suggestedOpeningQuote(
  settlementTarget: number,
  currentOffer = 0,
): number {
  const safeTarget = Math.max(0, settlementTarget);
  const safeOffer = Math.max(0, currentOffer);
  const cushion = Math.max(MIN_QUOTE_CUSHION, safeTarget * QUOTE_CUSHION_PCT);
  const rounded =
    Math.ceil((safeTarget + cushion) / QUOTE_ROUNDING_INCREMENT) *
    QUOTE_ROUNDING_INCREMENT;
  return roundMoney(Math.max(safeOffer, rounded));
}

/** Compare a posted broker offer with the contribution-profit rate bands. */
export function compareOfferToThresholds(
  currentOffer: number,
  rates: { minimum: number; good: number; great: number },
): OfferComparison {
  const offer = Math.max(0, currentOffer);
  const differenceVsGreat = roundMoney(offer - rates.great);

  if (offer >= rates.great) {
    return {
      position: "GREAT",
      differenceVsGreat,
      settlementTarget: null,
      suggestedCounteroffer: null,
    };
  }

  const position: OfferPosition = offer >= rates.good
    ? "GOOD"
    : offer >= rates.minimum
      ? "MARGINAL"
      : "BELOW_MINIMUM";
  const settlementTarget = position === "GOOD" ? rates.great : rates.good;

  return {
    position,
    differenceVsGreat,
    settlementTarget,
    suggestedCounteroffer: suggestedOpeningQuote(settlementTarget, offer),
  };
}

/**
 * Solving for the rate.
 *
 * Dispatch and factoring are usually a cut of the rate, so they move when the
 * rate moves and cannot simply be added to the cost. With
 *
 *   f = dispatch% + factoring%   (as a fraction)
 *   C = fuel + tolls + other + overhead + any FLAT dispatch/factoring fees
 *   P = target profit = target profit per mile x total miles
 *
 * the rate has to satisfy   R - C - fR = P,   which gives
 *
 *   R = (C + P) / (1 - f)
 *
 * That single expression produces every tier below; only P changes.
 */
export function calculateTargetRate(
  input: TargetRateInput,
  thresholds: RatingThresholds,
): TargetRate {
  const loadedMiles = Math.max(0, input.loadedMiles);
  const totalMiles = loadedMiles + Math.max(0, input.deadheadMiles);
  const gallons = input.mpg > 0 ? totalMiles / input.mpg : 0;
  const fuelCost = roundMoney(gallons * Math.max(0, input.fuelPrice));
  const tolls = roundMoney(Math.max(0, input.tolls));
  const otherCost = roundMoney(Math.max(0, input.otherCost));
  const overhead = roundMoney(totalMiles * Math.max(0, input.overheadPerMile));
  const debtService = roundMoney(totalMiles * Math.max(0, input.debtServicePerMile ?? 0));

  const flatFees = roundMoney(
    (input.dispatchMode === "AMOUNT" ? Math.max(0, input.dispatchValue) : 0) +
      (input.factoringMode === "AMOUNT" ? Math.max(0, input.factoringValue) : 0),
  );
  const grossFeeRate =
    (input.dispatchMode === "PCT" ? Math.max(0, input.dispatchValue) : 0) / 100 +
    (input.factoringMode === "PCT" ? Math.max(0, input.factoringValue) : 0) / 100;

  const directFixedCost = roundMoney(fuelCost + tolls + otherCost + flatFees);
  const fixedTripCost = roundMoney(directFixedCost + overhead);
  const impossible = grossFeeRate >= 1;

  const rateFor = (costAndProfit: number): number => {
    if (impossible || totalMiles <= 0) return 0;
    return roundMoney(costAndProfit / (1 - grossFeeRate));
  };

  const tierDefs: {
    key: RateTier["key"];
    label: string;
    ppm: number;
    numerator: number;
    description: string;
  }[] = [
    {
      key: "operatingBreakeven",
      label: "Operating break-even",
      ppm: 0,
      numerator: directFixedCost + overhead,
      description: "Covers direct trip costs and allocated operating costs. Excludes debt service.",
    },
    {
      key: "cashBreakeven",
      label: "Cash break-even",
      ppm: 0,
      numerator: directFixedCost + overhead + debtService,
      description: "Covers operating break-even plus the separately allocated debt-service burden.",
    },
    {
      key: "minimum",
      label: "Minimum acceptable",
      ppm: thresholds.marginal,
      numerator: directFixedCost + thresholds.marginal * totalMiles,
      description: `Contributes $${thresholds.marginal.toFixed(2)}/mi after direct trip costs -- your MARGINAL floor.`,
    },
    {
      key: "good",
      label: "Good rate",
      ppm: thresholds.good,
      numerator: directFixedCost + thresholds.good * totalMiles,
      description: `Contributes $${thresholds.good.toFixed(2)}/mi after direct trip costs -- rates as GOOD.`,
    },
    {
      key: "great",
      label: "Great rate",
      ppm: thresholds.great,
      numerator: directFixedCost + thresholds.great * totalMiles,
      description: `Contributes $${thresholds.great.toFixed(2)}/mi after direct trip costs -- rates as GREAT.`,
    },
    {
      key: "target",
      label: "Your target",
      ppm: input.targetProfitPerMile,
      numerator: directFixedCost + overhead + input.targetProfitPerMile * totalMiles,
      description: `Clears the $${input.targetProfitPerMile.toFixed(2)}/mi you asked for.`,
    },
  ];

  return {
    totalMiles,
    gallons,
    fuelCost,
    tolls,
    otherCost,
    overhead,
    debtService,
    directFixedCost,
    fixedTripCost,
    costPerMile: div(fixedTripCost, totalMiles),
    grossFeeRate,
    flatFees,
    directCostBreakEven: rateFor(directFixedCost),
    tiers: tierDefs.map((tier) => {
      const rate = rateFor(tier.numerator);
      return {
        key: tier.key,
        label: tier.label,
        profitPerMile: tier.ppm,
        rate,
        ratePerLoadedMile: div(rate, loadedMiles),
        description: tier.description,
      } satisfies RateTier;
    }),
    valid: totalMiles > 0 && input.mpg > 0 && !impossible,
    impossible,
  };
}

File: mobile/Sources/OnRoadBooks/Features/Calculator/LoadCalculator.swift
import Foundation

/// Ported from the web app's `src/lib/finance/load-calculator.ts` — same
/// cost model, same comment explaining WHY (kept verbatim below), so the
/// two never quietly drift apart.
///
///   Total miles      = loaded + deadhead to pickup      (always both)
///   Fuel             = total miles / MPG x price per gallon
///   Tolls            = entered
///   Dispatch         = percent of gross, or a flat amount
///   Factoring        = percent of gross, or a flat amount
///   Other            = entered
///   Overhead         = total miles x overhead cost per mile
///
/// `overheadPerMile` is meant to be the truck's own historical cost per
/// mile with fuel, tolls, dispatch and factoring already removed — those
/// four are entered explicitly above, and charging them twice is the
/// classic way a load calculator lies to you.
enum FeeMode: String, CaseIterable, Identifiable {
    case percent = "%", amount = "$"
    var id: String { rawValue }
}

struct CostLine: Identifiable {
    let id = UUID()
    let label: String
    let amount: Double
    let note: String?
}

struct LoadEstimate {
    let totalMiles: Double
    let grossPerTotalMile: Double
    let deadheadPct: Double
    let fuelCost: Double
    let tolls: Double
    let dispatch: Double
    let factoring: Double
    let otherCost: Double
    let overhead: Double
    let debtService: Double
    let tripCost: Double
    let totalCost: Double
    let contributionProfit: Double
    let contributionProfitPerMile: Double
    let contributionMargin: Double
    let profit: Double
    let profitPerMile: Double
    let profitMargin: Double
    let lines: [CostLine]
    let rating: LoadRating
    let valid: Bool
}

struct TargetRates {
    let directCostBreakEven: Double
    let operatingBreakEven: Double
    let cashBreakEven: Double
    let minimum: Double
    let good: Double
    let great: Double
    let customTarget: Double
    let openingQuote: Double
    let totalMiles: Double
    let impossible: Bool
}

enum OfferPosition {
    case great, good, marginal, belowMinimum
}

struct OfferComparison {
    let position: OfferPosition
    let differenceVsGreat: Double
    let settlementTarget: Double?
    let suggestedCounteroffer: Double?
}

enum LoadCalculatorMath {
    private static func roundMoney(_ v: Double) -> Double { (v * 100).rounded() / 100 }
    private static func div(_ a: Double, _ b: Double) -> Double { b > 0 ? a / b : 0 }

    private static func feeAmount(mode: FeeMode, value: Double, gross: Double) -> Double {
        let safe = value.isFinite && value > 0 ? value : 0
        return mode == .percent ? roundMoney(gross * (safe / 100)) : roundMoney(safe)
    }

    /// Exactly `rateLoad` from the web app: the account's own saved bands, and
    /// nothing else.
    ///
    /// This used to carry a "deadhead above 40% drops the band" rule that the
    /// web app does not have — invented here, and it meant the same load could
    /// be GOOD on a laptop and MARGINAL on a phone. Deadhead is already inside
    /// profit per mile, because every cost is charged across total miles.
    private static func rate(profitPerMile: Double, thresholds: RatingThresholds) -> LoadRating {
        if profitPerMile >= thresholds.great { return .great }
        if profitPerMile >= thresholds.good { return .good }
        if profitPerMile >= thresholds.marginal { return .marginal }
        return .bad
    }

    static func evaluate(
        grossRate: Double,
        loadedMiles: Double,
        deadheadMiles: Double,
        fuelPrice: Double,
        mpg: Double,
        tolls: Double,
        dispatchMode: FeeMode,
        dispatchValue: Double,
        factoringMode: FeeMode,
        factoringValue: Double,
        otherCost: Double,
        overheadPerMile: Double,
        debtServicePerMile: Double,
        thresholds: RatingThresholds
    ) -> LoadEstimate {
        let loadedMiles = max(0, loadedMiles)
        let deadheadMiles = max(0, deadheadMiles)
        let totalMiles = loadedMiles + deadheadMiles
        let grossRate = max(0, grossRate)

        let gallons = mpg > 0 ? totalMiles / mpg : 0
        let fuelCost = roundMoney(gallons * max(0, fuelPrice))
        let tollsAmt = roundMoney(max(0, tolls))
        let dispatch = feeAmount(mode: dispatchMode, value: dispatchValue, gross: grossRate)
        let factoring = feeAmount(mode: factoringMode, value: factoringValue, gross: grossRate)
        let other = roundMoney(max(0, otherCost))
        let overhead = roundMoney(totalMiles * max(0, overheadPerMile))
        let debtService = roundMoney(totalMiles * max(0, debtServicePerMile))

        let tripCost = roundMoney(fuelCost + tollsAmt + dispatch + factoring + other)
        let totalCost = roundMoney(tripCost + overhead)
        let contributionProfit = roundMoney(grossRate - tripCost)
        let contributionProfitPerMile = div(contributionProfit, totalMiles)
        let contributionMargin = grossRate > 0 ? div(contributionProfit, grossRate) * 100 : 0
        let profit = roundMoney(contributionProfit - overhead)
        let profitPerMile = div(profit, totalMiles)
        let profitMargin = grossRate > 0 ? div(profit, grossRate) * 100 : 0
        let deadheadPct = div(deadheadMiles, totalMiles)

        let lines: [CostLine] = [
            CostLine(label: "Fuel", amount: fuelCost,
                     note: mpg > 0 ? String(format: "%.1f gal at $%.2f/gal, %.1f MPG", gallons, fuelPrice, mpg) : "Enter MPG to estimate fuel"),
            CostLine(label: "Tolls", amount: tollsAmt, note: nil),
            CostLine(label: "Dispatch", amount: dispatch, note: dispatchMode == .percent ? "\(Int(dispatchValue))% of gross" : "Flat fee"),
            CostLine(label: "Factoring", amount: factoring, note: factoringMode == .percent ? "\(Int(factoringValue))% of gross" : "Flat fee"),
            CostLine(label: "Other costs", amount: other, note: nil),
        ]

        return LoadEstimate(
            totalMiles: totalMiles,
            grossPerTotalMile: div(grossRate, totalMiles),
            deadheadPct: deadheadPct,
            fuelCost: fuelCost,
            tolls: tollsAmt,
            dispatch: dispatch,
            factoring: factoring,
            otherCost: other,
            overhead: overhead,
            debtService: debtService,
            tripCost: tripCost,
            totalCost: totalCost,
            contributionProfit: contributionProfit,
            contributionProfitPerMile: contributionProfitPerMile,
            contributionMargin: contributionMargin,
            profit: profit,
            profitPerMile: profitPerMile,
            profitMargin: profitMargin,
            lines: lines,
            rating: rate(profitPerMile: contributionProfitPerMile, thresholds: thresholds),
            valid: totalMiles > 0 && mpg > 0
        )
    }

    /// R = (C + P) / (1 - f) — solving for the rate that clears a target
    /// profit per mile, given fixed costs C and a combined dispatch+factoring
    /// fee rate f (see the header comment in the web source for the algebra).
    static func targetRate(
        loadedMiles: Double,
        deadheadMiles: Double,
        fuelPrice: Double,
        mpg: Double,
        tolls: Double,
        dispatchMode: FeeMode,
        dispatchValue: Double,
        factoringMode: FeeMode,
        factoringValue: Double,
        otherCost: Double,
        overheadPerMile: Double,
        debtServicePerMile: Double,
        thresholds: RatingThresholds,
        targetProfitPerMile: Double
    ) -> TargetRates {
        let totalMiles = max(0, loadedMiles) + max(0, deadheadMiles)
        let gallons = mpg > 0 ? totalMiles / mpg : 0
        let fuelCost = roundMoney(gallons * max(0, fuelPrice))
        let tollsAmt = roundMoney(max(0, tolls))
        let other = roundMoney(max(0, otherCost))
        let overhead = roundMoney(totalMiles * max(0, overheadPerMile))
        let debtService = roundMoney(totalMiles * max(0, debtServicePerMile))

        let flatFees =
            (dispatchMode == .amount ? max(0, dispatchValue) : 0) +
            (factoringMode == .amount ? max(0, factoringValue) : 0)
        let feeRate =
            (dispatchMode == .percent ? max(0, dispatchValue) : 0) / 100 +
            (factoringMode == .percent ? max(0, factoringValue) : 0) / 100

        let directCost = roundMoney(fuelCost + tollsAmt + other + flatFees)
        let fixedCost = roundMoney(directCost + overhead)
        let impossible = feeRate >= 1
        guard !impossible else {
            return TargetRates(
                directCostBreakEven: 0, operatingBreakEven: 0, cashBreakEven: 0,
                minimum: 0, good: 0, great: 0, customTarget: 0,
                openingQuote: 0, totalMiles: totalMiles, impossible: true
            )
        }

        func rateFor(_ amount: Double) -> Double { roundMoney(amount / (1 - feeRate)) }
        let directCostBreakEven = rateFor(directCost)
        let operatingBreakEven = rateFor(fixedCost)
        let cashBreakEven = rateFor(fixedCost + debtService)
        let minimum = rateFor(directCost + thresholds.marginal * totalMiles)
        let good = rateFor(directCost + thresholds.good * totalMiles)
        let great = rateFor(directCost + thresholds.great * totalMiles)
        let targetProfit = targetProfitPerMile * totalMiles
        let customTarget = rateFor(fixedCost + targetProfit)
        let openingQuote = suggestedOpeningQuote(settlementTarget: max(great, customTarget))
        return TargetRates(
            directCostBreakEven: directCostBreakEven,
            operatingBreakEven: operatingBreakEven,
            cashBreakEven: cashBreakEven,
            minimum: minimum,
            good: good,
            great: great,
            customTarget: customTarget,
            openingQuote: openingQuote,
            totalMiles: totalMiles,
            impossible: false
        )
    }

    static func suggestedOpeningQuote(settlementTarget: Double, currentOffer: Double = 0) -> Double {
        let safeTarget = max(0, settlementTarget)
        let cushion = max(25, safeTarget * 0.03)
        let rounded = ceil((safeTarget + cushion) / 25) * 25
        return roundMoney(max(max(0, currentOffer), rounded))
    }

    static func compareOffer(_ currentOffer: Double, rates: TargetRates) -> OfferComparison {
        let offer = max(0, currentOffer)
        let difference = roundMoney(offer - rates.great)
        if offer >= rates.great {
            return OfferComparison(
                position: .great, differenceVsGreat: difference,
                settlementTarget: nil, suggestedCounteroffer: nil
            )
        }
        let position: OfferPosition = offer >= rates.good
            ? .good
            : offer >= rates.minimum ? .marginal : .belowMinimum
        let settlementTarget = position == .good ? rates.great : rates.good
        return OfferComparison(
            position: position,
            differenceVsGreat: difference,
            settlementTarget: settlementTarget,
            suggestedCounteroffer: suggestedOpeningQuote(
                settlementTarget: settlementTarget, currentOffer: offer
            )
        )
    }
}
